pages/dashboard.js

```diff
@@ -0,0 +1,226 @@
+    backgroundColor: "transparent",
+  },
+  {
+    id: "1345",
+    content:
+      "ned or moved too fast. I think it’s been good and a successful program that we’ve gotten this far this fast. Remember though that we still think there’s a need for ongoing rate increases, a",
+    backgroundColor: "transparent",
+  },
+];
+
+const colorSet = [
+  { id: "5454", color: "transparent" },
+  { id: "87", color: "#E8E19C" },
+  { id: "987", color: "#A5E89C" },
+  { id: "986", color: "#9CD0E8" },
+  { id: "455", color: "#E194B8" },
+];
+const MAX_DROP_ITEMS = 4;
+
+export default function Dashboard() {
+  const [draggableItems, setDraggableItems] = useState(draglist);
+  const [remainDragItemCount, setRemainDragItemCount] = useState(
+    draglist.length
+  );
+
+  const [dropzones, setDropzones] = useState([
+    {
+      id: Math.random().toString(36).substring(2),
+    },
+    { text: "Drop here!" },
+  ]);
+
+  const [colorPalette, setColorPalette] = useState({});
+  const [currentDraggable, setCurrentDraggable] = useState({
+    id: "",
+  });
+
+  function openColorPalette(event, id) {
+    event.preventDefault();
+    setCurrentDraggable({ id: id });
+
+    //popup the color pallete at right mouse click
+    setColorPalette({
+      display: true,
+      posX: event.pageX + "px",
+      posY: event.pageY + "px",
+    });
+  }
+
+  //set color background for the clicked draggable item
+  function chooseColor(color) {
+    setDraggableItems((prevlist) =>
+      prevlist.map((item) =>
+        item.id === currentDraggable.id
+          ? { ...item, backgroundColor: color }
+          : item
+      )
+    );
+    setColorPalette((pl) => {
+      return { ...pl, display: false };
+    });
+  }
+
+  function drag(event) {
+    event.dataTransfer.setData("text", event.target.id);
+  }
+  function drop(event) {
+    event.preventDefault();
+    const dataID = event.dataTransfer.getData("text");
+
+    //only allow drop for an amount equal to MAX_DROP_ITEMS items
+    if (
+      event.target.id === "DraggableContainer" ||
+      event.target.children.length <= MAX_DROP_ITEMS
+    )
+      event.currentTarget.appendChild(
+        document.querySelector(`#${CSS.escape(dataID)}`)
+      );
+  }
+
+  function removeBoard(id) {
+    setDropzones((dropzones) => dropzones.filter((dz) => dz.id !== id));
+  }
+
+  useEffect(() => {
+    //close the color palette by clicking outside it
+    document.body.addEventListener("click", (event) => {
+      if (event.target.id !== "ColorPalette")
+        setColorPalette((pl) => {
+          return { ...pl, display: false };
+        });
+    });
+    return () => {
+      document.body.removeEventListener("click", (event) => {
+        if (event.target.id !== "ColorPalette")
+          setColorPalette((pl) => {
+            return { ...pl, display: false };
+          });
+      });
+    };
```

Comment:
Very good!

If you wanted to slim down the code further you could write a function for the callback and hand down this function in addEventListener and removeEventListener. This is optional, you choose :)

Reply:
i know i was so lazy. I m having many problems with the other PR it drove me quite mad 🙈 😅. But I will do it!